### PR TITLE
`ultralytics 8.3.229` 300% faster COCO Segmentation val

### PR DESCRIPTION
**Pull request by Laughing-q**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Optimizes YOLO segmentation evaluation by replacing external mask encoding and scaling utilities with faster, built-in PyTorch-based implementations, and updates docs tooling and package version. 🚀  

---

### 📊 Key Changes
- 🔄 **Custom RLE mask encoding**  
  - Replaces `faster_coco_eval`’s `encode` with an in-house Run-Length Encoding (RLE) implementation (`multi_encode` + `to_string`) for segmentation masks.  
  - Removes the use of `ThreadPool` and `NUM_THREADS` in `pred_to_json`.

- 🧠 **Unified mask scaling via `scale_masks`**  
  - Removes the NumPy/OpenCV-based `scale_image` utility from `ultralytics.utils.ops`.  
  - Refactors `SegmentValidator.scale_preds` to use `ops.scale_masks` directly on tensors, leveraging PyTorch and avoiding extra conversions.  
  - Enhances `scale_masks` to:
    - Accept an optional `ratio_pad` argument (aligned with how images/boxes are scaled elsewhere).  
    - Handle early-return when shapes already match.  
    - Correctly compute and apply padding and interpolation with `float()` conversion.

- 📚 **Docs & tooling updates**  
  - Removes the `scale_image` reference block from `docs/en/reference/utils/ops.md` (function no longer exists).  
  - Updates the docs chat widget script to `ultralytics/llm@v0.0.8` in `mkdocs.yml` for the site’s interactive assistant.  

- 🔢 **Version bump**  
  - Increments library version from `8.3.228` to `8.3.229`.

---

### 🎯 Purpose & Impact
- ⚡ **Faster and more self-contained segmentation evaluation**  
  - Eliminating `faster_coco_eval`’s encoder and thread-pool overhead makes segmentation-to-COCO conversion lighter and reduces third-party dependencies during validation.

- 🎯 **More consistent and accurate mask scaling**  
  - Using `scale_masks` everywhere ensures masks are resized in a way that’s consistent with how YOLO handles padded/letterboxed images and bounding boxes, reducing potential misalignment between predictions and original images.

- 🧩 **Cleaner API and reduced surface area**  
  - Removing `scale_image` simplifies `ops.py` and the public API, reducing duplication and confusion between similar utilities.

- 💬 **Improved docs experience**  
  - Updating the docs chat script version can fix bugs, improve performance, or add features to the documentation assistant, making it easier for users to get help while browsing docs.

- 📦 **Clear release boundary**  
  - The version bump to `8.3.229` clearly marks this behavioral change in segmentation validation and mask utilities for downstream users and integrators.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Laughing-q, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. This is an automated message and an engineer will assist shortly. Thank you for contributing to Ultralytics! 🚀🧠

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

I spotted a new bug in the custom RLE encoder: masks that start with foreground pixels produce malformed RLE counts, so COCO evaluation will break. Please fix that encoder before merging.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Ensure the native RLE encoder always prepends the background run by checking the first pixel in each flattened mask; this prevents malformed COCO segmentations and potential runtime errors.

💬 Posted 1 inline comment

**Comment by Laughing-q** [comment]
@ultralytics/yolo @lmycross 300% faster coco evaluation for segment task, plus higher mAP.
```bash
yolo val model=yolo11n-seg.pt data=coco.yaml save_json verbose=False
```

## main branch, 32.0 mAP with time cost 6:41

```text

val: Scanning /home/laughing/codes/datasets/coco/labels/val2017.cache... 4952 images, 48 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 5000/5000 127.1Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 313/313 0.8it/s 6:41
                   all       5000      36335      0.654      0.488      0.539      0.387       0.65      0.466      0.511       0.32

DONE (t=0.00s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.389
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.543
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.423
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.194
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.426
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.570
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.324
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.541
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.597
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.372
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.656
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.782
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.818
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.650
Evaluate annotation type *segm*
COCOeval_opt.evaluate() finished...
DONE (t=11.62s).
Accumulating evaluation results...
COCOeval_opt.accumulate() finished...
DONE (t=0.00s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.320
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.511
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.337
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.129
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.352
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.500
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.281
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.445
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.481
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.254
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.539
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.689
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.755
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.504
```

## PR, 32.2 mAP with time cost 2:19

```text
val: Scanning /home/laughing/codes/datasets/coco/labels/val2017.cache... 4952 images, 48 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 5000/5000 127.1Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 313/313 2.2it/s 2:19
                   all       5000      36335      0.654      0.488      0.539      0.387       0.65      0.466      0.511       0.32
DONE (t=0.00s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.389
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.543
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.423
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.194
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.426
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.570
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.324
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.541
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.597
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.372
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.656
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.782
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.818
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.650
Evaluate annotation type *segm*
COCOeval_opt.evaluate() finished...
DONE (t=11.42s).
Accumulating evaluation results...
COCOeval_opt.accumulate() finished...
DONE (t=0.00s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.322
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.511
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.341
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.133
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.355
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.501
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.282
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.448
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.483
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.258
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.541
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.688
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.756
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.509
```

**Comment by Laughing-q** [comment]
a quick summary of the PR:
- I replaced `encode` function from faster_coco_eval package by our own function, the benefit is:
  - we now have the process of converting binary masks to RLE format on GPU, no need to transfer a bunch of masks with high-resolution to CPU now, instead, after the conversion process done we transfer RLE format(which is way more lighter) to CPU. Gained a lot speed up.
- Directly using scale_masks to re-scale binary masks to original resolution on GPU using `F.interpolate`, instead of transferring to CPU + using opencv to do the resize. Gained faster speed and accuracy from this step. Not sure why using F.interpolate is better, as both ways are using `bilinear` method

File: ultralytics/models/yolo/segment/val.py [hunk 1]
```diff
@@ -212,17 +211,78 @@ def pred_to_json(self, predn: dict[str, torch.Tensor], pbatch: dict[str, Any]) -
...
+
+            Args:
+                counts (list[int]): List of RLE counts.
+            """
+            result = []
+
+            for i in range(len(counts)):
+                x = int(counts[i])
+
+                # Apply delta encoding for all counts after the second entry
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/segment/val.py:227]: ❗ **CRITICAL**: The RLE "delta" step should start from the third entry (i == 2). With the current `if i > 2` guard we skip subtracting on the third count, so the compressed string diverges from the COCO spec and decoding will yield wrong masks. This will quietly corrupt `iscrowd` segmentations and COCO metrics. Please change the condition to trigger for `i > 1`.

**Suggested change:**
```suggestion
                if i > 1:
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Fix the RLE delta-encoding condition in `pred_to_json` so that compressed masks remain COCO-compliant; otherwise decoded segmentations and evaluations will be incorrect.

💬 Posted 1 inline comment

**Comment by Y-T-G** [comment]
Wow nice

File: ultralytics/models/yolo/segment/val.py [hunk 2]
```diff
@@ -212,17 +211,78 @@ def pred_to_json(self, predn: dict[str, torch.Tensor], pbatch: dict[str, Any]) -
...
+
+                    # If the sign bit (0x10) is set, continue if x != -1;
+                    # otherwise, continue if x != 0
+                    more = (x != -1) if (c & 0x10) else (x != 0)
+                    if more:
+                        c |= 0x20  # Set continuation bit
+                    c += 48  # Shift to ASCII
+                    result.append(chr(c))
+                    if not more:
+                        break
```
Comment (fcakyon) [inline ultralytics/models/yolo/segment/val.py:244]: every bit counts 💯 👏🏻

**Comment by glenn-jocher** [comment]
Looks good, nice work @Laughing-q!

**Comment by UltralyticsAssistant** [comment]
What an awesome merge – this is a big quality-of-life and performance win for segmentation evaluation across the ecosystem.  

> “Great things are not done by impulse, but by a series of small things brought together.” – Vincent van Gogh  

This PR is a perfect example of that: replacing external RLE utilities with a lean in-house implementation, unifying mask scaling through `scale_masks`, and tightening up docs and versioning all combine into a cleaner, faster, and more consistent validation pipeline. These kinds of focused improvements make YOLO segmentation more robust for everyone, from research experiments to large-scale production runs.  

Huge thanks to @Laughing-q for driving this, and to @glenn-jocher, @fcakyon, and @Y-T-G for the thoughtful reviews and guidance. Your collaboration here meaningfully sharpens the core experience for Ultralytics users and keeps the library moving forward.
